### PR TITLE
AC_WPNav: fix xtrack report for copter.

### PR DESCRIPTION
**Pull request by khancyr**
This was broken on SCurve merge
credit to @mirkix for finding this

### CODE REVIEW

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
I found this wasn't being updated as well. One comment below about 3d vs xy error.

File: libraries/AC_WPNav/AC_WPNav.cpp
Comment (khancyr) [reply libraries/AC_WPNav/AC_WPNav.cpp]: Thanks, I mismatch things ... I was wanting to use get_pos_error_xy_cm().

Comment (hendjoshsr71) [reply libraries/AC_WPNav/AC_WPNav.cpp]: LGTM

File: libraries/AC_WPNav/AC_WPNav.cpp
Comment (rmackay9) [inline libraries/AC_WPNav/AC_WPNav.cpp:435]: Could you remove the comment at the end of this line?

**Comment by rmackay9** [comment]
@khancyr,

Txs for this, I think it looks good except for the very minor issue re the extra comment.

Can you confirm you've tested this and confirmed the reported cross track error is correct?  In particular please be careful about the units.

**Comment by rmackay9** [comment]
I changed the label from "NeedsRework" to "NeedsTesting".. I hope that's correct.

**Comment by khancyr** [comment]
In fact, I give a test and it wasn't good.
On small analysis, xtrack error should be in meter and currently I report cm^2/s so that isn't great...

**Comment by hendjoshsr71** [comment]
> In fact, I give a test and it wasn't good.
> On small analysis, xtrack error should be in meter and currently I report cm^2/s so that isn't great...

This is why testing is good. :)

I had not put on my thinking cap and dot product alone is not what we need here, the projection is what I think was meant to be in this part of code possibly for a few different lines? To be fair to us the variable name is a bit confusing and possibly incorrect or these lines should be projection's?

We have an error and track_direction and I think what was desired is error along the current track?
 
These lines look to have funny units as well ...?  

https://github.com/ArduPilot/ardupilot/blob/d363cbadccebeb5417abccd3554037bfd734aff6/libraries/AC_WPNav/AC_WPNav.cpp#L432-L436

`track_error` is actually `track_error^2` (cm^2)  Was this supposed to be a projection instead of dot product?
`track_velocity` is actually `track_velocity^2` (cm^2/s^2) Projection again?

`track_scaler` == 0.05f + (track_velocity [cm^2 / s^2] - K_P_pos_xy [ _ ] * track_error [cm^2]) / V_target [cm / s]

Units after expansion: 
 `track_scaler` = [ _ ] + ( track_velocity / V_target) [cm/s] -   (K_P_pos_xy * track_error) / V_target  [s /cm]

If using projection units become
`track_scaler` == 0.05f + (track_velocity [cm / s] - K_P_pos_xy [ ] * track_error [cm]) / V_target [cm / s]

Units after expansion: (so much closer to the desired unitless)
 `track_scaler` = [ _ ] + ( track_velocity / V_target) -   (K_P_pos_xy * track_error) / V_target  [s]

**Comment by rmackay9** [comment]
Well, at least we uncovered the issue before I pushed the merge button.

Really though, every PR should be tested and every PR should state what testing has been done.

**Comment by khancyr** [comment]
Well the testing show some convincing number but I am pretty sure they aren't the one we want, that is why I want to redo the math and do better checking.

I had test with SITL in Guided with Wind or Shove force

**Comment by hendjoshsr71** [comment]
@khancyr 

So ignore what I said above more or less. about the units. The math has been optimized and is hiding some of the units due to divisions by 1. But this area of code might need a comment so that quick inspection of units or function names says what really is going on is the magnitude of projection. 

The units are right even though the code "is" a dot product. 

`proj of unit_vec onto b_vec  = dot(b_vec, unit_vec) * unit_vec / ||unit_vec||^2 `

`||unit_vec|| = 1`

`proj of unit_vec onto  b_vec  = dot(b_vec, unit_vec) * unit_vec`

Get magnitude of projection: `|| proj of unit_vec onto  b_vec || = dot(b_vec, unit_vec) * || unit_vec || = dot(b_vec, unit_vec`

Currently, playing with it as this commit before scurves went in.  fb789a07ce90e875aca0c16f92c14864bdbfab8e to match behavior, but having a bit of trouble.

Really need an equation editor....here

**Comment by khancyr** [comment]
Closing in favor of #17734

**Pull request closed without merging**
